Opt in

torotil opened the pull request:
Provide generic opt-in functionality (especially for phone and post).

* Create a new webform component that has similar features as the newsletter component, but without the newsletter list integration.
* Record opt-ins in activities.
* Add contact fields in campaignion starterkit.
* …

torotil committed:
Copy over needed functionality from campaignion_newsletters.

torotil committed:
Implement opt-out checkboxes.

torotil committed:
Consistently use the same values accross the module.

torotil committed:
Fix labels that still mention newsletters.

torotil committed:
Add basic tests for the component.

torotil committed:
Add Opt-in fields to the palette.

torotil committed:
Add opt-in data to activities.

- Adds the concept of opt-in communication channels to differentiate
  between post, email and phone opt-ins.

torotil committed:
Remove the generic opt-in from the palette.

torotil committed:
Support for newsletter components.

torotil committed:
Refactor opt-in recording and add tests.

torotil committed:
Add BooleanOptIn importer.

torotil committed:
Add BooleanField exporter.

katzmo commented on campaignion_opt_in/tests/OptInRecordTest.php:
```diff
@@ -0,0 +1,84 @@
+  /**
+   * Test a webform submission with an opt_in and a newsletter component.
+   */
+  public function testTwoOptIuns() {
```
typo

katzmo commented on campaignion_opt_in/campaignion_opt_in.module:
```diff
@@ -47,9 +48,9 @@ function campaignion_opt_in_webform_conditional_form($node) {
-          'no' => t('Opt-out'),
+          Values::OPT_IN => t('Opt-in'),
+          Values::NO_CHANGE => t('No change'),
+          Values::OPT_OUT => t('Opt-out'),
```
conditionals need not selected option too.

torotil replied:
Hm, thinking about this again and I don’t think it’s actually needed after all.

torotil committed:
Migrate from direct_mail_newsletter to post_opt_in.

katzmo committed:
[campaignion_opt_in] fix: don't set non-existing default value

Radios can only be submitted with a value that corresponds to any of the
radio options or 0.

katzmo committed:
[campaignion_opt_in] re-define inverted checkboxes

Definition: "inverted" = doing something to do nothing:
* unchecked means opt-in
* checked means no change

katzmo committed:
[campaignion_opt_in] Consolidate values with campaignion_newsletters (draft)

torotil committed:
Implement JS conditionals for the inverted checkbox.

torotil committed:
Add explicit #parents to the channel setting.

torotil committed:
Implement opt-in recording with prefixed values.

torotil committed:
Remove unused API function.

torotil committed:
Save both a flag and the full component value in the activities.

torotil committed:
Update for prefixed values.

katzmo commented on campaignion_opt_in/campaignion_opt_in.conditionals.js:
```diff
@@ -10,8 +10,10 @@
+      var isInverted = checkbox.value === 'no-change';
+      var uncheckedValue = isInverted ? 'opt-in' : 'no-change';
+      var value = checkbox.checked ? checkbox.value : uncheckedValue;
+      return ruleValue === 'checkbox:' + value;
```
Brilliant idea!

katzmo commented on campaignion_newsletters/src/Component.php:
```diff
@@ -77,21 +77,6 @@ public function submit($email, WebformSubmission $s) {
-    $value = ValuePrefix::remove($value);
-    return $value == 'opt-in';
-  }
-
```
We should make a decision if we find this useful or not and consolidate with https://github.com/moreonion/campaignion/pull/111 at some point in the future.

torotil replied:
Yes, I agree.

katzmo commented on campaignion_opt_in/src/OptInRecordFactory.php:
```diff
@@ -40,7 +40,8 @@ public function recordOptIn($component, $values) {
       db_insert('campaignion_opt_in')->fields([
         'activity_id' => $this->activity->activity_id,
-        'operation' => reset($values),
+        'operation' => $values == Values::OPT_IN ? 1 : 0,
```
Should be `$value == Values::OPT_IN` (without `s`)!

katzmo replied:
Pushed a fix.

katzmo commented on campaignion_opt_in/campaignion_opt_in.install:
```diff
@@ -44,6 +50,7 @@ function campaignion_opt_in_schema() {
     'indexes' => [
       'activity' => ['activity_id'],
       'channel' => ['channel'],
+      'operation' => ['operation'],
```
Good idea!

katzmo commented on campaignion_form_builder/campaignion_form_builder.type_opt_in.inc:
```diff
@@ -0,0 +1,42 @@
+    $element = [
+      'title' => $component['name'],
+      'weight' => $weight,
+      'unique' => TRUE,
```
Not sure if we want it to be unique, right now, multiple newsletter fields are possible (using campaignion_newsletters)?

torotil replied:
Our current interpretation of phone and post opt-ins is that they are contact-wide. So it wouldn’t make much sense to have more than one field per form, right?

katzmo replied:
OK.

katzmo reviewed (changes requested):
$values → $value needs fixing, rest looks great!
(Still not completely happy with default values, but I think it’s ok for now ;))

katzmo committed:
[campaignion_opt_in] fix typo
